# <center> HEART DISEASE PREDICTION </center>

Heart disease prediction using 3 machine learning models.
1. Support Vector Machine 
2. K Nearset Neighbour 
3. ANN Multilayer Perceptron

## DATA DESCRIPTION

South Africa Heart Disease Dataset
Source:<br>
https://web.stanford.edu/~hastie/ElemStatLearn//data.html<br>
https://www.openml.org/d/1498


A retrospective sample of males in a heart-disease high-risk region of the Western Cape, South Africa. There are roughly two controls per case of CHD. Many of the CHD positive men have undergone blood pressure reduction treatment and other programs to reduce their risk factors after their CHD event. In some cases the measurements were made after these treatments. These data are taken from a larger dataset, described in Rousseauw et al, 1983, South African Medical
Journal. 

Attributes:

1. sbp systolic blood pressure 
2. tobacco cumulative tobacco (kg) 
3. ldl low densiity lipoprotein cholesterol 
4. adiposity 
5. famhist family history of heart disease (Present, Absent) 
6. typea type-A behavior 
7. obesity 
8. alcohol current alcohol consumption 
9. age age at onset 
10. chd response, coronary heart disease 

In [27]:
# Importing primary libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib notebook

### DATA PREPROCESSING  

In [28]:
# Reading the data 
data = pd.read_csv('heart.csv')
data.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

In [29]:
# Setting up the column
column = ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach','exang', 'oldpeak', 'slope', 'ca', 'thal', 'target']

In [30]:
data.columns=column

In [31]:
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [32]:
data.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

Describes the statistical data insights

In [33]:
# Checking for any missing values 
data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

The values shown are 0, therefore no missing values in the DataFrame

Data preprocessing is complete,
implementing the models

In [34]:
X = data.iloc[:,0:13].values#not including target
y = data.iloc[:,13].values

In [35]:
#normalize the my whole data into o to 1 range

from sklearn.preprocessing import StandardScaler
X_std = StandardScaler().fit_transform(X)

dfNorm = pd.DataFrame(X_std, index=data.index, columns=data.columns[0:13])
dfNorm.head()

age       sex        cp  trestbps      chol       fbs   restecg  \
0  0.952197  0.681005  1.973123  0.763956 -0.256334  2.394438 -1.005832   
1 -1.915313  0.681005  1.002577 -0.092738  0.072199 -0.417635  0.898962   
2 -1.474158 -1.468418  0.032031 -0.092738 -0.816773 -0.417635 -1.005832   
3  0.180175  0.681005  0.032031 -0.663867 -0.198357 -0.417635  0.898962   
4  0.290464 -1.468418 -0.938515 -0.663867  2.082050 -0.417635  0.898962   

    thalach     exang   oldpeak     slope        ca      thal  
0  0.015443 -0.696631  1.087338 -2.274579 -0.714429 -2.148873  
1  1.633471 -0.696631  2.122573 -2.274579 -0.714429 -0.512922  
2  0.977514 -0.696631  0.310912  0.976352 -0.714429 -0.512922  
3  1.239897 -0.696631 -0.206705  0.976352 -0.714429 -0.512922  
4  0.583939  1.435481 -0.379244  0.976352 -0.714429 -0.512922

In [36]:
# # add non-feature target column to dataframe
dfNorm['target'] = data['target']


X = dfNorm.iloc[:,0:13].values
y = dfNorm.iloc[:,13].values
dfNorm.head(10)

age       sex        cp  trestbps      chol       fbs   restecg  \
0  0.952197  0.681005  1.973123  0.763956 -0.256334  2.394438 -1.005832   
1 -1.915313  0.681005  1.002577 -0.092738  0.072199 -0.417635  0.898962   
2 -1.474158 -1.468418  0.032031 -0.092738 -0.816773 -0.417635 -1.005832   
3  0.180175  0.681005  0.032031 -0.663867 -0.198357 -0.417635  0.898962   
4  0.290464 -1.468418 -0.938515 -0.663867  2.082050 -0.417635  0.898962   
5  0.290464  0.681005 -0.938515  0.478391 -1.048678 -0.417635  0.898962   
6  0.180175 -1.468418  0.032031  0.478391  0.922521 -0.417635 -1.005832   
7 -1.143291  0.681005  0.032031 -0.663867  0.323431 -0.417635  0.898962   
8 -0.260980  0.681005  1.002577  2.306004 -0.913400  2.394438  0.898962   
9  0.290464  0.681005  1.002577  1.049520 -1.512490 -0.417635  0.898962   

    thalach     exang   oldpeak     slope        ca      thal  target  
0  0.015443 -0.696631  1.087338 -2.274579 -0.714429 -2.148873       1  
1  1.633471 -0.696631  2.122573 -2.274579 -0.714429 -0.512922       1  
2  0.977514 -0.696631  0.310912  0.976352 -0.714429 -0.512922       1  
3  1.239897 -0.696631 -0.206705  0.976352 -0.714429 -0.512922       1  
4  0.583939  1.435481 -0.379244  0.976352 -0.714429 -0.512922       1  
5 -0.072018 -0.696631 -0.551783 -0.649113 -0.714429 -2.148873       1  
6  0.146634 -0.696631  0.224643 -0.649113 -0.714429 -0.512922       1  
7  1.021244 -0.696631 -0.896862  0.976352 -0.714429  1.123029       1  
8  0.540209 -0.696631 -0.465514  0.976352 -0.714429  1.123029       1  
9  1.064975 -0.696631  0.483451  0.976352 -0.714429 -0.512922       1

In [37]:
#split into testing and training dataset
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

X_train.shape, y_train.shape, X_test.shape , y_test.shape

((212, 13), (212,), (91, 13), (91,))

# <center> MODEL 1 - Support Vector Machine</center>

A Support Vector Machine (SVM) is a discriminative classifier formally defined by a separating hyperplane. In other words, given labeled training data (supervised learning), the algorithm outputs an optimal hyperplane which categorizes new examples.<br>
Description : https://en.wikipedia.org/wiki/Support_vector_machine <br>
Kernel used : Linear<br>
Degree =3<br>
Gamma = Auto

In [12]:
from sklearn import svm
svm_clf = svm.SVC(kernel ='linear')

In [13]:
svm_clf.fit(X_train,y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='linear', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False)

In [14]:
y_pred_svm =svm_clf.predict(X_test)

In [15]:
y_pred_svm

array([0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0,
       0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0,
       0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1,
       1, 0, 0], dtype=int64)

In [16]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm_svm = confusion_matrix(y_test, y_pred_svm)
cm_svm

array([[31, 13],
       [ 5, 42]], dtype=int64)

In [17]:
from sklearn.metrics import accuracy_score
svm_result = accuracy_score(y_test,y_pred_svm)
print("Accuracy :",svm_result)

Accuracy : 0.8021978021978022


In [18]:
recall_svm = cm_svm[0][0]/(cm_svm[0][0] + cm_svm[0][1])
precision_svm = cm_svm[0][0]/(cm_svm[0][0]+cm_svm[1][1])
recall_svm,precision_svm

(0.7045454545454546, 0.4246575342465753)

## Accuracy : 0.802
## Recall : 0.704
## Precision : 0.4246

# <center> MODEL 2 - K Nearest Neighbour</center>

In pattern recognition, the k-nearest neighbors algorithm (k-NN) is a non-parametric method used for classification and regression. In both cases, the input consists of the k closest training examples in the feature space. ... In k-NN classification, the output is a class membership.<br>
Description : https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm<br>
Neighbors =5<br>
N jobs = No of cores ,-1<br>
Leaf size =60<br>
Algorithm = Brute

In [19]:
from sklearn.neighbors import KNeighborsClassifier
knn_clf = KNeighborsClassifier(n_neighbors =5,n_jobs = -1,leaf_size = 60,algorithm='brute')

In [20]:
knn_clf.fit(X_train,y_train)

KNeighborsClassifier(algorithm='brute', leaf_size=60, metric='minkowski',
                     metric_params=None, n_jobs=-1, n_neighbors=5, p=2,
                     weights='uniform')

In [21]:
y_pred_knn = knn_clf.predict(X_test)
y_pred_knn

array([0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0,
       0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0,
       1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 0], dtype=int64)

In [22]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm_knn = confusion_matrix(y_test, y_pred_knn)
cm_knn

array([[32, 12],
       [ 4, 43]], dtype=int64)

In [23]:
knn_result = accuracy_score(y_test,y_pred_knn)
knn_result

0.8241758241758241

In [24]:
recall_knn = cm_knn[0][0]/(cm_knn[0][0] + cm_knn[0][1])
precision_knn = cm_knn[0][0]/(cm_knn[0][0]+cm_knn[1][1])
recall_knn,precision_knn

(0.7272727272727273, 0.4266666666666667)

## Accuracy : 0.645
## Recall : 0.816
## Precision : 0.816

### <center>Cross Validation with KFold</center>

Cross Validation recovers from the random state difference by diving the complete data set in K parts, and iteratively learns from data.<br>
Folds =10

In [26]:
from sklearn.model_selection import KFold

fold_list =[]
accuracy_list =[]
def run_kfold(knn_clf):
    kf = KFold(n_folds=10)
    outcomes = []
    fold = 0
    for train_index, test_index in kf:
        fold += 1
        X_train, X_test = X_all.values[train_index], X_all.values[test_index]
        y_train, y_test = y_all.values[train_index], y_all.values[test_index]
        knn_clf.fit(X_train, y_train)
        predictions = knn_clf.predict(X_test)
        accuracy = accuracy_score(y_test, predictions)
        outcomes.append(accuracy)
        print("Fold {0} accuracy: {1}".format(fold, accuracy))
        fold_list.append(fold)
        accuracy_list.append(accuracy)
        mean_outcome = np.mean(outcomes)
        print("Mean Accuracy: {0}".format(mean_outcome)) 

run_kfold(knn_clf)
    

TypeError: __init__() got an unexpected keyword argument 'n_folds'

Box plot of the iterative learning by K Fold in K-Nearest Neighbors

In [ ]:
fold_df =pd.DataFrame({'Fold':fold_list,'Accuracy':accuracy_list}).set_index('Fold')
fold_df.plot(kind='box',figsize=(4,6),color=color)

### Fold 10 accuracy: 0.7586206896551724 
### Mean Accuracy: 0.6501149425287356

# MODEL 3 - ANN Multilayer Perceptron Classifier

### <center> ANN implementation using Scikit learn library </center>

Input nodes : 9<br>
Hidden layers :2<br>
Each hidden Layer Hold 14 neuron<br>
Output layer : 2

In [38]:
from sklearn.metrics import make_scorer, accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.neural_network import MLPClassifier

'''
Multi-layer Perceptron classifier.

This model optimizes the log-loss function using LBFGS or stochastic
gradient descent.
'''

'\nMulti-layer Perceptron classifier.\n\nThis model optimizes the log-loss function using LBFGS or stochastic\ngradient descent.\n'

In [53]:
ann_clf = MLPClassifier()

#when i used acitcation function not relu accuray was not goind to excedd the 76.98 when i change with it to relu it's acccuraccy increase to 80% yahooooo

#The solver for weight optimization.
'''

    - 'lbfgs' is an optimizer in the family of quasi-Newton methods.

    - 'sgd' refers to stochastic gradient descent.

    - 'adam' refers to a stochastic gradient-based optimizer proposed
      by Kingma, Diederik, and Jimmy Ba
'''


# now alpha as we know in back propagation we use it
'''
alpha : float, optional, default 0.0001
    L2 penalty (regularization term) parameter.
'''


#echos
'''
max_iter : int, optional, default 200
    Maximum number of iterations
'''


# now see the shuffle

'''
shuffle : bool, optional, default True
    Whether to shuffle samples in each iteration. Only used when
    solver='sgd' or 'adam'.
    
'''
#Parameters
parameters = {'solver': ['adam'],#lbfgs
              'activation':['relu'],
             'alpha':[1e-4],
             'hidden_layer_sizes':(13,17,17,2),   # 9 input, 14-14 neuron in 2 layers,1 output layer
              'random_state': [1]
             }

In [54]:
# Type of scoring to compare parameter combos 
acc_scorer = make_scorer(accuracy_score)

# Run grid search 
grid_obj = GridSearchCV(ann_clf, parameters, scoring=acc_scorer)
grid_obj = grid_obj.fit(X_train, y_train)

# Pick the best combination of parameters
ann_clf = grid_obj.best_estimator_

C:\Users\Server\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
C:\Users\Server\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
C:\Users\Server\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
C:\Users\Server\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, Conv

In [55]:
# Fit the best algorithm to the data 
ann_clf.fit(X_train, y_train)

C:\Users\Server\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=13, learning_rate='constant',
              learning_rate_init=0.001, max_iter=200, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=1, shuffle=True, solver='adam', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

In [56]:
y_pred_ann = ann_clf.predict(X_test)

In [57]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm_ann = confusion_matrix(y_test, y_pred_ann)
cm_ann

array([[32, 12],
       [ 6, 41]], dtype=int64)

In [58]:
ann_result = accuracy_score(y_test,y_pred_ann)
ann_result

0.8021978021978022

In [ ]:
recall_ann = cm_ann[0][0]/(cm_ann[0][0] + cm_ann[0][1])
precision_ann = cm_ann[0][0]/(cm_ann[0][0]+cm_ann[1][1])
recall_ann,precision_ann

## Accuracy : 0.763
## Recall : 0.866
## Precision : 0.732

### <center> ANN implementation using keras library </center>

Sequential Model
Input node :13
Hidden layer :2
Each layer hold 6 neuron

In [64]:
import keras
from keras.models import Sequential
from keras.layers import Dense

# Initialising the ANN
classifier = Sequential()

# Adding the input layer and the first hidden layer
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu', input_dim = 13))

# Adding the second hidden layer
#classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu'))

# Adding the output layer
classifier.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))

# Compiling the ANN
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

# Fitting the ANN to the Training set
history = classifier.fit(X_train, y_train, batch_size = 10, nb_epoch = 100)

# Part 3 - Making the predictions and evaluating the model

# Predicting the Test set results
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm_annk = confusion_matrix(y_test, y_pred)

C:\Users\Server\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=13, units=6, kernel_initializer="uniform")`
  if __name__ == '__main__':
C:\Users\Server\Anaconda3\lib\site-packages\ipykernel_launcher.py:15: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="uniform")`
  from ipykernel import kernelapp as app
C:\Users\Server\Anaconda3\lib\site-packages\ipykernel_launcher.py:21: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Epoch 1/100
212/212 [==============================] - 0s 1ms/step - loss: 0.6911 - accuracy: 0.7689
Epoch 2/100
212/212 [==============================] - 0s 184us/step - loss: 0.6854 - accuracy: 0.8349
Epoch 3/100
212/212 [==============================] - 0s 205us/step - loss: 0.6741 - accuracy: 0.8396
Epoch 4/100
212/212 [==============================] - 0s 185us/step - loss: 0.6568 - accuracy: 0.8538
Epoch 5/100
212/212 [==============================] - 0s 208us/step - loss: 0.6330 - accuracy: 0.8538
Epoch 6/100
212/212 [==============================] - 0s 162us/step - loss: 0.6034 - accuracy: 0.8538
Epoch 7/100
212/212 [==============================] - 0s 210us/step - loss: 0.5723 - accuracy: 0.8491
Epoch 8/100
212/212 [==============================] - 0s 160us/step - loss: 0.5412 - accuracy: 0.8491
Epoch 9/100
212/212 [==============================] - 0s 188us/step - loss: 0.5101 - accuracy: 0.8538
Epoch 10/100
212/212 [==============================] - 0s 185us/step - los

212/212 [==============================] - 0s 190us/step - loss: 0.3007 - accuracy: 0.8821
Epoch 80/100
212/212 [==============================] - 0s 162us/step - loss: 0.3003 - accuracy: 0.8821
Epoch 81/100
212/212 [==============================] - 0s 166us/step - loss: 0.2995 - accuracy: 0.8821
Epoch 82/100
212/212 [==============================] - 0s 188us/step - loss: 0.2994 - accuracy: 0.8821
Epoch 83/100
212/212 [==============================] - 0s 188us/step - loss: 0.2991 - accuracy: 0.8774
Epoch 84/100
212/212 [==============================] - 0s 195us/step - loss: 0.2988 - accuracy: 0.8821
Epoch 85/100
212/212 [==============================] - 0s 179us/step - loss: 0.2982 - accuracy: 0.8821
Epoch 86/100
212/212 [==============================] - 0s 192us/step - loss: 0.2981 - accuracy: 0.8774
Epoch 87/100
212/212 [==============================] - 0s 179us/step - loss: 0.2976 - accuracy: 0.8868
Epoch 88/100
212/212 [==============================] - 0s 169us/step - loss:

In [60]:
import matplotlib.pyplot as plt
%matplotlib inline
# Model accuracy

plt.plot(history.history['accuracy_score'])
plt.plot(history.history['val_acc'])
plt.title('Model Accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'test'])
plt.show()

KeyError: 'accuracy_score'

Confusion matrix obtained from keras NN 

In [ ]:
cm_annk

In [ ]:
recall_annk = cm_annk[0][0]/(cm_annk[0][0] + cm_annk[0][1])
precision_annk = cm_annk[0][0]/(cm_annk[0][0]+cm_annk[1][1])
recall_annk,precision_annk

## Accuracy  ~  0.7317
## Recall : 0.783
## Precision : 0.701

## Comparison of models

In [ ]:
results ={'Accuracy': [svm_result*100,knn_result*100,ann_result*100],
          'Recall': [recall_svm*100,recall_knn*100,recall_ann*100],
          'Precision': [precision_svm*100,precision_knn*100,precision_ann*100]}
index = ['SVM','KNN','ANN']

In [ ]:
results =pd.DataFrame(results,index=index)

Bar Plot of three model implemented<br>
X axis : Models<br>
Y axis : Values of Accuracy, Precision,Recall

In [ ]:
fig =results.plot(kind='bar',title='Comaprison of models',figsize =(9,9)).get_figure()
fig.savefig('Final Result.png')


Zoomed view having lower y limit 50, higher y limit 100, for better view 

In [ ]:
fig =results.plot(kind='bar',title='Comaprison of models',figsize =(6,6),ylim=[50,100]).get_figure()
fig.savefig('image.png')

Subplot of each measure in between the models

In [ ]:
results.plot(subplots=True,kind ='bar',figsize=(4,10))

In [ ]:
Saving graph images
DataFrame
fig = class_counts.plot().get_figure()
fig.savefig('image.png')

In [ ]:
# create X and Y datasets for training
from sklearn import model_selection

X = np.array(data.drop(['target'], 1))
y = np.array(data['target'])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, test_size = 0.2)

In [ ]:
# convert the data to categorical labels
from keras.utils.np_utils import to_categorical

Y_train = to_categorical(y_train, num_classes=None)
Y_test = to_categorical(y_test, num_classes=None)
print (Y_train.shape)
print (Y_train[:10])

In [ ]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam

# define a function to build the keras model
def create_model():
    # create model
    model = Sequential()
    model.add(Dense(16, input_dim=13, kernel_initializer='normal', activation='relu'))
    model.add(Dense(8, kernel_initializer='normal', activation='relu'))
    model.add(Dense(2, activation='softmax'))
    
    # compile model
    adam = Adam(lr=0.001)
    model.compile(loss='categorical_crossentropy', optimizer=adam, metrics=['accuracy'])
    return model

model = create_model()

print(model.summary())

In [ ]:
# fit the model to the training data
history=model.fit(X_train, y_train, validation_data=(X_test, y_test),epochs=200, batch_size=10, verbose = 10)